media_driver/linux/common/ddi/media_ddi_render_target_table.cpp

```diff
@@ -0,0 +1,362 @@
+//!           required to register the new render target, VA_STATUS_ERROR_NOT_ENOUGH_BUFFER if it is
+//!           impossible to register the new render target because none are marked as inactive.
+//!
+//!
+
+VAStatus DDI_CODEC_RENDER_TARGET_TABLE::RegisterRTSurface(VASurfaceID id)
+{
+    if (id == VA_INVALID_ID)
+    {
+        DDI_ASSERTMESSAGE("Invalid VASurfaceID in RegisterRTSurfaces");
+        return VA_STATUS_ERROR_INVALID_PARAMETER;
+    }
+
+    if (IsRegistered(id))
+    {
+        return VA_STATUS_SUCCESS;
+    }
+
+    if (!m_free_index_pool.empty())
+    {
+        RTTableIdx idx = m_free_index_pool.front();
+        m_free_index_pool.pop_front();
+
+        DDI_CODEC_RENDER_TARGET_INFO info;
+        info.FrameIdx = idx;
+        info.RTState = RT_STATE_ACTIVE_IN_CURFRAME;
+
+        m_va_to_rt_map[id] = info;
+        return VA_STATUS_SUCCESS;
+    }
+    else
+    {
+        using RTPair = std::pair<VASurfaceID, DDI_CODEC_RENDER_TARGET_INFO>;
+        using RTPairIterator = std::map<VASurfaceID, DDI_CODEC_RENDER_TARGET_INFO>::iterator;
+        std::vector<RTPairIterator> inactive_rt_vec;
+        inactive_rt_vec.reserve(m_va_to_rt_map.size());
+
+        for (RTPairIterator it = m_va_to_rt_map.begin(); it != m_va_to_rt_map.end(); it++)
+        {
+            if (it->second.RTState == RT_STATE_INACTIVE)
+            {
+                inactive_rt_vec.push_back(it);
+            }
+        }
+
+        if (inactive_rt_vec.empty())
+        {
+            return VA_STATUS_ERROR_NOT_ENOUGH_BUFFER;
+        }
+
+        // Evict the inactive render target with the smallest FrameIdx
+        RTPairIterator rt_to_evict_iter = *std::min_element(inactive_rt_vec.begin(), inactive_rt_vec.end(),
+                [](const RTPairIterator& lhs,
+                   const RTPairIterator& rhs)
+                   {return lhs->second.FrameIdx < rhs->second.FrameIdx;});
+        RTTableIdx freed_idx = rt_to_evict_iter->second.FrameIdx;
+
+        m_va_to_rt_map.erase(rt_to_evict_iter);
+        m_va_to_rt_map[id].FrameIdx = freed_idx;
+        m_va_to_rt_map[id].RTState = RT_STATE_ACTIVE_IN_CURFRAME;
+
+        return VA_STATUS_ERROR_MAX_NUM_EXCEEDED;
+    }
+}
+
+//!
+//! \brief    Unregister Render Target Surface
+//! \details  Unregister surface in render target table
+//!
+//! \param    [in] id
+//!           VASurfaceID for the render target to be unregistered
+//!
+//! \return   VAStatus
+//!           VA_STATUS_SUCCESS if successful, VA_STATUS_ERROR_INVALID_PARAMETER if
+//!           no such surface is registered
+//!
+VAStatus DDI_CODEC_RENDER_TARGET_TABLE::UnRegisterRTSurface(VASurfaceID id)
+{
+    if (!IsRegistered(id))
+    {
+        DDI_VERBOSEMESSAGE("UnRegisterRTSurface: render target was not registered in the RTtbl!");
+        return VA_STATUS_ERROR_INVALID_PARAMETER;
+    }
+
+    m_free_index_pool.push_front(m_va_to_rt_map[id].FrameIdx);
```

Comment:
u push/pop elements from the same end of container (front) - what is intention to use `deque` instead of `vector` + `push_back/pop_back`

Reply:
A queue-like structure looks more natural here; besides, this makes it easier to implement other surface eviction policies, if need be.